### PR TITLE
Fixing Database.mergeHistory for EOL and error time steps

### LINKED ISSUE #1960: Further issues found with the EOL being written to the Database
@albeanth and I identified two places where the ARMI codebase does not support the new `EOL` time point being written to the database:

1. `Database3.genTimeSteps()`:

https://github.com/terrapower/armi/blob/8b5163a8b61ebe704419d2d77067bfcf8018f328/armi/bookkeeping/db/database3.py#L612

Returns `(cycle, node)`, but if the `EOL` is present in the DB, it just returns the last time node twice.

2. `Database3.load()` takes in `cycle` and `node` as integers, thus not allowing us to load an EOL timestep:

https://github.com/terrapower/armi/blob/8b5163a8b61ebe704419d2d77067bfcf8018f328/armi/bookkeeping/db/database3.py#L702

### PR DESCRIPTION
**Pull request by john-science**
## What is the change? Why is it being made?

This is a follow up to https://github.com/terrapower/armi/pull/2607. That solution solve the problem, but opened up another problem in `Database.mergeHistory()`, which also needs to support "error" and "EOL" time nodes.

**NOTE**: I also went through all the rest of the uses of `Database.genTimeSteps()` in ARMI, and they are not similarly broken.

close https://github.com/terrapower/armi/issues/1960

## SCR Information

 Change Type: fixes

One-Sentence Rationale: MergeHistory should not fail if there are special error or EOL time steps in the DB.

One-line Impact on Requirements: NA

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/bookkeeping/db/tests/test_database.py [hunk 1]
```diff
@@ -1032,15 +1032,72 @@ def test_genTimeStepGroups(self):
         # remove the fake H5 file or the DB cleanup in ARMI's context.py will panic
         db.h5db = None
 
+    def test_mergeHistory(self):
+        """Test some edge cases of Database.mergeHistory.
+
+        This test uses some mock database and H5 file tooling, so we can more easily test edge cases in the ........
```
Comment (john-science) [inline armi/bookkeeping/db/tests/test_database.py]:
```suggestion
        This test uses some mock database and H5 file tooling, so we can more easily test edge cases.
```

File: armi/bookkeeping/db/tests/test_database.py [hunk 2]
```diff
@@ -1032,15 +1032,72 @@ def test_genTimeStepGroups(self):
...
+
+        # test 2: should grab the first time node and the error time node
+        results = []
+        db.mergeHistory(MockDB([(0, 0), (0, 0, "error"), (0, 1)]), 0, 1)
+        self.assertListEqual(results, [(0, 0), (0, 0, "error")])
+
+        # test 3: should grab all 5 time nodes, including the EOL
+        results = []
+        db.mergeHistory(MockDB([(0, 0), (0, 1), (1, 0), (1, 1), (1, 1, "EOL")]), 2, 4)
+        self.assertListEqual(results, [(0, 0), (0, 1), (1, 0), (1, 1), (1, 1, "EOL")])
```
Comment (drewj-tp) [inline armi/bookkeeping/db/tests/test_database.py:1090]: Odd but I guess it works to show we can handle an EOL state too

File: armi/bookkeeping/db/tests/test_database.py [hunk 3]
```diff
@@ -1032,15 +1032,72 @@ def test_genTimeStepGroups(self):
...
+
+        # test 1: should grab the first 4 of 6 time nodes
+        results = []
+        db.mergeHistory(MockDB([(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1)]), 2, 0)
+        self.assertListEqual(results, [(0, 0), (0, 1), (1, 0), (1, 1)])
+
+        # test 2: should grab the first time node and the error time node
+        results = []
+        db.mergeHistory(MockDB([(0, 0), (0, 0, "error"), (0, 1)]), 0, 1)
+        self.assertListEqual(results, [(0, 0), (0, 0, "error")])
```
Comment (drewj-tp) [inline armi/bookkeeping/db/tests/test_database.py:1085]: Would it be possible to have `c00n00error` and `c00n01`? There was an error fatal enough to write to a database but we somehow wrote the next statepoint?

Maybe more realistically is someone had an error and wants to restart prior to that error.

**Review by drewj-tp — COMMENTED** [review]
The mocking / optimization is almost too much to know if this is really going to do what we want.

I guess the alternative would be for every one of these small steps to use something like `unittest.mock.patch on
- `self.db.copy` to know that we copied the right things
- `inputDB.genTimeSteps` and `inputDB.genTimeStepGroups`

which doesn't feel better. IDK I'm rambling here. 

The changes to `database.py` look promising. Please run this through downstream testing as well

File: armi/bookkeeping/db/tests/test_database.py (hunk 2 above)
Comment (john-science) [reply armi/bookkeeping/db/tests/test_database.py:1090]: LOL

Yeah, I'm calling that out in the unit testing because I want to show we are explicitly handling everything: error and EOL.

The other option was to support neither, but the conversation was this was the more conservative option.

#shrug

I bet we don't need either of those. But no one wanted to risk it.

File: armi/bookkeeping/db/tests/test_database.py (hunk 3 above)
Comment (john-science) [reply armi/bookkeeping/db/tests/test_database.py:1085]: I mean, yeah, I think dropping error and EOL time steps seems valid.

I think, in reality, people probably don't care about the "error" time steps enough that they want to merge them in. But @opotowsky suggested this was the more conservative option.  The thought goes that no one will complain about having too much data in their output files.

I can flip that and drop EOL and error timesteps.  That's not hard to do, or anything. #shrug

Comment (opotowsky) [reply armi/bookkeeping/db/tests/test_database.py:1085]: I don't have a strong opinion, we'd just want to make sure it wouldn't hurt anyone's workflow by changing prior behavior.

**Review by opotowsky — APPROVED** [review]
I see the other review material but I'm throwing in my 2c approval. 

My take: The unit tests don't make sense bc the behavior doesn't make sense, and John was just trying to retain the nonsensical behavior. We should just make a ticket to clean up the behavior

**Comment by john-science** [comment]
> We should just make a ticket to clean up the behavior

If we can find some real world users of the restart cases, we could get some ground truth on what the best functionality here should be.
